Fix comparison functions and add unit tests

gabriele-ciccotelli opened the pull request:
### Description
This PR fixes the implementation of the comparison functions `WeightedDeltaCompare` and `KeypointCompare`.
The previous versions of these functions could return incorrect results when the compared fields were equal, potentially causing inconsistent ordering or crashes when used with `qsort`.

### Changes
- Corrected the implementation of `WeightedDeltaCompare` and `KeypointCompare` so that they now return 0 when the compared fields are equal.
- Added unit tests to verify the correct behavior of both functions.

### Additional changes for cross-platform (non-Android) support:
To ensure the tests and build work correctly on both Android and non-Android platforms (such as Linux), the following changes were made:
- Wrapped the `#include <android/log.h>` directive with `#ifdef __ANDROID__` in relevant files.
- Added stub definitions for Android logging macros and functions in `logging.h` for non-Android platforms.
- Moved the definition of the `WeightedDelta` struct to the header file (`frame_pair.h`) so it is visible to tests and other files.
- Declared the comparison functions `WeightedDeltaCompare` and `KeypointCompare` in their respective header files (`frame_pair.h` and `keypoint_detector.h`).

### How to Build and Run the Tests on Android
1. Build the test binary for Android ARM64:
```bash
bazel-7.4.1 build --config=android_arm64 //tensorflow/tools/android/test/jni/object_tracking:object_tracking_test
```
2. Push the binary to your Android device:
```bash
adb push <PATH_TO_BINARY>/object_tracking_test /data/local/tmp/object_tracking_test_manual
adb shell chmod +x /data/local/tmp/object_tracking_test_manual
```
4. Run the tests on the device:
```bash
adb shell /data/local/tmp/object_tracking_test_manual
```

### How to Build and Run the Tests on Linux
1. Build the test binary for Linux:
```bash
bazel-7.4.1 build //tensorflow/tools/android/test/jni/object_tracking:object_tracking_test
```
2. Run the tests:
```bash
./bazel-bin/tensorflow/tools/android/test/jni/object_tracking/object_tracking_test
```

### Test Results
All tests pass successfully:

```text
Running main() from gmock_main.cc
[==========] Running 8 tests from 2 test suites.
[----------] Global test environment set-up.
[----------] 4 tests from WeightedDeltaCompareTest
[ RUN      ] WeightedDeltaCompareTest.ReturnsZeroWhenEqual
[       OK ] WeightedDeltaCompareTest.ReturnsZeroWhenEqual (0 ms)
[ RUN      ] WeightedDeltaCompareTest.ReturnsPositiveWhenALessThanB
[       OK ] WeightedDeltaCompareTest.ReturnsPositiveWhenALessThanB (0 ms)
[ RUN      ] WeightedDeltaCompareTest.ReturnsNegativeWhenAGreaterThanB
[       OK ] WeightedDeltaCompareTest.ReturnsNegativeWhenAGreaterThanB (0 ms)
[ RUN      ] WeightedDeltaCompareTest.QSortStableOrdering
[       OK ] WeightedDeltaCompareTest.QSortStableOrdering (0 ms)
[----------] 4 tests from WeightedDeltaCompareTest (0 ms total)

[----------] 4 tests from KeypointCompareTest
[ RUN      ] KeypointCompareTest.ReturnsZeroWhenEqual
[       OK ] KeypointCompareTest.ReturnsZeroWhenEqual (0 ms)
[ RUN      ] KeypointCompareTest.ReturnsPositiveWhenALessThanB
[       OK ] KeypointCompareTest.ReturnsPositiveWhenALessThanB (0 ms)
[ RUN      ] KeypointCompareTest.ReturnsNegativeWhenAGreaterThanB
[       OK ] KeypointCompareTest.ReturnsNegativeWhenAGreaterThanB (0 ms)
[ RUN      ] KeypointCompareTest.QSortStableOrdering
[       OK ] KeypointCompareTest.QSortStableOrdering (0 ms)
[----------] 4 tests from KeypointCompareTest (0 ms total)

[----------] Global test environment tear-down
[==========] 8 tests from 2 test suites ran. (0 ms total)
[  PASSED  ] 8 tests.
```

Resolves #91989

gabriele-ciccotelli committed:
Fix comparison functions and add unit tests

- Corrected the implementation of  `WeightedDeltaCompare()` and `KeypointCompare()` so that they now return 0 when the compared fields are equal;
- Added unit tests to verify the correct behavior of both functions.

gabriele-ciccotelli committed:
Fix comparison functions and add unit tests

- Corrected the implementation of  `WeightedDeltaCompare()` and `KeypointCompare()` so that they now return 0 when the compared fields are equal;
- Added unit tests to verify the correct behavior of both functions.

mihaimaruseac reviewed (dismissed):
I don't think these belong here.

mihaimaruseac commented on tensorflow/tools/android/test/jni/object_tracking/frame_pair.cc:
```diff
@@ -192,10 +192,22 @@ struct WeightedDelta {
 inline int WeightedDeltaCompare(const void* const a, const void* const b) {
-  return (reinterpret_cast<const WeightedDelta*>(a)->delta -
-          reinterpret_cast<const WeightedDelta*>(b)->delta) <= 0 ? 1 : -1;
+    float delta_a = reinterpret_cast<const WeightedDelta*>(a)->delta;
+    float delta_b = reinterpret_cast<const WeightedDelta*>(b)->delta;
+    if (delta_a < delta_b) return -1;
+    if (delta_a > delta_b) return 1;
+    return 0;
```
This changes the order of the results. Previously if `a->delta == 1` and `b->delta == 2` the code would return 1 whereas now it would return -1.

Recommend to skip the clamping to 1/-1 and instead return just the difference.

mihaimaruseac commented on tensorflow/tools/android/test/jni/object_tracking/keypoint_detector.cc:
```diff
@@ -67,9 +67,22 @@ void KeypointDetector::ScoreKeypoints(const ImageData& image_data,
 inline int KeypointCompare(const void* const a, const void* const b) {
-  return (reinterpret_cast<const Keypoint*>(a)->score_ -
-          reinterpret_cast<const Keypoint*>(b)->score_) <= 0 ? 1 : -1;
+  float score_a = reinterpret_cast<const Keypoint*>(a)->score_;
+  float score_b = reinterpret_cast<const Keypoint*>(b)->score_;
+  if (score_a < score_b) return -1;
+  if (score_a > score_b) return 1;
+  return 0;
```
Same comments here

mihaimaruseac commented on tensorflow/tools/android/test/jni/object_tracking/object_tracking_test.cc:
```diff
@@ -0,0 +1,115 @@
+    EXPECT_FLOAT_EQ(arr[2].score_, 2.0f);
+}
+
+}
```
Please ensure the file ends with a newline character (not an empty line)

gabriele-ciccotelli commented:
Thank you @mihaimaruseac for your feedback!

Regarding the newline character, I will address it right away.

For the two comparison functions, the reason I chose to use an `if-else` structure instead of directly returning the difference is that, according to the official documentation for the `qsort` function, the comparison function must return an integer (`int`). If the difference between two `float` values is returned directly, the result will be implicitly converted to an `int`, truncating the value.

For example, if the difference between the two values is `0.5`, casting to `int` would return `0`, so the two elements would be considered equal even though they are not.

In any case, please let me know if you would still prefer that I return the difference directly, and I will make that change as well if needed.

Thanks again.

mihaimaruseac commented:
Let's go with the if-else code then. Thank you for the context.

gabriele-ciccotelli committed:
Merge branch 'tensorflow:master' into master

gabriele-ciccotelli committed:
Invert comparison functions return values, update tests and comments

- Invert return values in WeightedDeltaCompare() and KeypointCompare() (return 1 if a < b, -1 if a > b, 0 if equal)
- Update function comments to reflect new behavior
- Add newline at end of test file
- Update unit tests to match new return values
- All tests pass successfully

gabriele-ciccotelli committed:
Merge branch 'tensorflow:master' into master

gabriele-ciccotelli committed:
Merge branch 'master' into fix-comparison-functions

gabriele-ciccotelli committed:
Merge branch 'tensorflow:master' into master

gabriele-ciccotelli committed:
Merge branch 'master' into fix-comparison-functions

gabriele-ciccotelli committed:
Invert comparison functions return values, update tests and comments

- Invert return values in WeightedDeltaCompare() and KeypointCompare() (return 1 if a < b, -1 if a > b, 0 if equal);
- Update function comments to reflect new behavior;
- Add newline at end of test file;
- Update unit tests to match new sorting order;
- All tests pass successfully.

mihaimaruseac reviewed (dismissed):
Thank you!

gabriele-ciccotelli committed:
Added #ifdef to logging.h to prevent errors generated by the test suite

gabriele-ciccotelli commented:
Hi,

I have analyzed the logs generated by the failed tests.

### Py+CPP Test Suite
For the following tests:
- Presubmit - Linux Arm64 CPU Build Tests only - Py+CPP Test Suite;
- Presubmit - Linux x86 CPU - Py+CPP Test Suite;

The build fails with the following error:
```text
fatal error: 'android/log.h' file not found
```
I believe this happens because the tests are executed on Linux, while the `android/log.h` header is only available on Android platforms.
I noticed that this `#include` appears in `logging.h`, `logging.cc`, and `object_tracker_jni.cc`.
To address this, if you agree, I can wrap the include with:
```cc
#ifdef __ANDROID__
#include <android/log.h>
#endif
```
I think that this should prevent the error when testing on non-Android platforms.

### Code Check
This other test fails because of the BUILD format. I can reorder the elements as the log suggests.

Let me know if you have any other suggestions or if this approach is correct.

mihaimaruseac commented:
Hi, both changes seem good to me.

This is still pending internal review.

gabriele-ciccotelli committed:
Added #ifdef to files including android/log.h and reordered elements of BUILD file

gabriele-ciccotelli committed:
Merge branch 'tensorflow:master' into master

gabriele-ciccotelli committed:
Merge branch 'master' into fix-comparison-functions

gabriele-ciccotelli committed:
Added #ifdef to files including android/log.h and reordered elements of BUILD file to prevent errors generated by the test suite

mihaimaruseac commented on tensorflow/tools/android/test/jni/object_tracking/BUILD:
```diff
@@ -0,0 +1,75 @@
+)
+
+# Definition of the test for the object tracking library
+cc_test(
```
Can you change this to `tf_native_cc_binary` instead? This is to make sure all dependencies are properly tracked in all environments.

gabriele-ciccotelli committed:
Fix header and stub issues for cross-platform build and test

- Moved WeightedDelta struct definition to frame_pair.h for visibility in tests and other files
- Declared WeightedDeltaCompare and KeypointCompare functions in their respective header files
- Added stub definitions for Android logging macros and functions in logging.h to support non-Android platforms
- Verified successful build and test on both Linux and Android
- Changed test rule from cc_test to tf_native_cc_binary in BUILD file

gabriele-ciccotelli committed:
Merge branch 'tensorflow:master' into master

gabriele-ciccotelli committed:
Merge branch 'master' into fix-comparison-functions

gabriele-ciccotelli committed:
Fix header and stub issues for cross-platform build and test

- Moved `WeightedDelta` struct definition to `frame_pair.h` for visibility in tests and other files;
- Declared `WeightedDeltaCompare` and `KeypointCompare` functions in their respective header files;
- Added stub definitions for Android logging macros and functions in `logging.h` to support non-Android platforms;
- Verified successful build and test on both Linux and Android;
- Changed test rule from `cc_test` to `tf_native_cc_binary` in BUILD file.

gabriele-ciccotelli committed:
Add load statement for tf_native_cc_binary in BUILD file

Added the required load("//tensorflow:tensorflow.bzl", "tf_native_cc_binary") statement at the top of the BUILD file.

gabriele-ciccotelli committed:
Merge branch 'tensorflow:master' into master

gabriele-ciccotelli committed:
Merge branch 'master' into fix-comparison-functions

gabriele-ciccotelli committed:
Add load statement for tf_native_cc_binary in BUILD file

Added the required load("//tensorflow:tensorflow.bzl", "tf_native_cc_binary") statement at the top of the BUILD file.

mihaimaruseac commented on tensorflow/tools/android/test/jni/object_tracking/BUILD:
```diff
@@ -0,0 +1,77 @@
+    deps = [
+        ":object_tracking_lib",
+        "@com_google_googletest//:gtest_main",
+    ],
```
```suggestion
    ],
    testonly = True,
```

mihaimaruseac committed:
Add testonly attribute

gabriele-ciccotelli committed:
Updated BUILD file

Updated BUILD file including "testonly = True"

gabriele-ciccotelli committed:
Merge branch 'tensorflow:master' into master

gabriele-ciccotelli committed:
Merge branch 'master' into fix-comparison-functions

gabriele-ciccotelli committed:
Updated BUILD file

Updated BUILD file including `testonly = True` option.

gabriele-ciccotelli committed:
Moved the attribute ‘testonly = True’ to the top as suggested by Code Check

gabriele-ciccotelli committed:
Merge branch 'tensorflow:master' into master

gabriele-ciccotelli committed:
Merge branch 'master' into fix-comparison-functions

gabriele-ciccotelli committed:
Moved the attribute ‘testonly = True’ to the top as suggested by Code Check

gabriele-ciccotelli committed:
fix(BUILD): replace tf_native_cc_binary with cc_binary

Adds load statements and updates object_tracking/BUILD to use standard cc_binary, fixing compatibility with newer Android NDK versions

gabriele-ciccotelli committed:
style(BUILD): sort load statements

Invert the order of loads to pass the presubmit code check

keerthanakadiri commented:
Hence, closing this PR due to inactivity. Please feel free to raise a new PR if you'd like to revisit these changes later.
